### ISSUE TITLE
Which cert does kubelet used actually?

### ISSUE THREAD

**Issue by stricklandye** [kubernetes/kubernetes#120820]
### What happened?

Our cluster recently encountered issues with certificate renewal, a master node  failed to renew certificate from apiserver  becuase there is out-dated base64 encoded kubelet.conf I think that's why kubelet cannot renew `kubelet-client-current.pem`.

### What did you expect to happen?

cert in kubelet.conf should point to `/var/lib/kublet/pki/kubelet-client-current.pem` after TLS bootstrap after finished cluster initialization instead of a base64 encoded cert.

### How can we reproduce it (as minimally and precisely as possible)?

No need to reproduce.

### Anything else we need to know?

According to the [doc](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/#workflow-when-using-kubeadm-join), base64 encoded cert should only used in kubelet start phase, but what if both cert provided in kubelet.conf and `/var/lib/kubelet/pki/kubelet-client-current.pem`, which cert does kubelet used for communicate with apiserver acutally ? 
Also the [warning](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#:~:text=rotation%20fails.-,Warning,-%3A) says base64 encoed cert only appears prior to 1.17 which makes confused because our cluster is 1.18

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:50:46Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```
</details>

### Cloud provider

<details>
bare-metalcluster built by ourselves
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here
Linux 4.9.0-12-amd64 #1 SMP Debian 4.9.210-1 (2020-01-20) x86_64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by pranav-pandey0804** [comment]
> which cert does kubelet used for communicate with apiserver actually ?

i think the kubelet prefers the current certificate file over the base64 encoded certificate, as stated in [this document](https://kubernetes.io/docs/tasks/tls/certificate-rotation/).

**Comment by stricklandye** [comment]
> > which cert does kubelet used for communicate with apiserver actually ?
> 
> i think the kubelet prefers the current certificate file over the base64 encoded certificate, as stated in [this document](https://kubernetes.io/docs/tasks/tls/certificate-rotation/).

Thanks for response. I think kubelet use the base64 encoded cert in kubelet.conf acqure a new certificate from apiserver and use `kubelet-client-current.pem` to communicate with apisever,right ?

**Comment by pranav-pandey0804** [comment]
@stricklandye You are welcome! 
Yes, you are correct. The `kubelet` uses the base64 encoded certificate in **kubelet.conf** to acquire a new certificate from the `apiserver`, and then uses the **kubelet-client-current.pem** file to communicate with the `apiserver`. This is how the `kubelet` performs **TLS bootstrapping**, which is a process of requesting a signed certificate from the apiserver and storing it in a local file for future use. The **kubelet.conf** file is updated with the path to the local file, so that the `kubelet` can use it for subsequent connections.

**Comment by pranav-pandey0804** [comment]
you can refer to these resources to understand this better :

- [TLS Bootstrapping | Kubernetes](https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping/): This is a document that explains how the kubelet performs TLS bootstrapping, which is the process of requesting a signed certificate from the apiserver and storing it in a local file for future use.
- [Issue #581](https://github.com/kubernetes/kubeadm/issues/581): This is an issue on GitHub that discusses the problem of certificate renewal in Kubernetes clusters created with kubeadm, and some possible solutions to fix it.

**Comment by pranav-pandey0804** [comment]
> Another question confuses me is that doc says since 1.18 cert in kubelet.conf should point to cert in /var/lib/kubelet/pki/kubelet-client-current.pem, but why is this an exception?

This is a good question. The documentation says that since 1.18, the cert in **kubelet.conf** should point to the cert in **/var/lib/kubelet/pki/kubelet-client-current.pem**, which is the file that contains the rotated kubelet client certificate. However, this is not always the case, as there are some exceptions:

- One exception is when you use `kubeadm` to create a cluster with an external CA (certificate authority). In this case, `kubeadm` will not generate a **kubelet.conf** file for you, and you have to create it manually. You also have to provide the base64 encoded certificate and key in the **kubelet.conf** file. you can refer to [this](https://stackoverflow.com/questions/64850658/how-to-add-certificates-to-kube-config-file) and [this](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#:~:text=External%20CA%20mode&text=If%20all%20other%20certificates%20and,the%20CA%20certificate%20and%20key.).

- Another exception is when you use `kubeadm` to upgrade a cluster from a version prior to 1.17 to a version 1.18 or later. In this case, `kubeadm` will not update the **kubelet.conf** file for you, and you have to modify it manually to point to the rotated certificate file, as explained  [here](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/).

These exceptions are due to some limitations of `kubeadm` and the way it handles **certificate management**. 
I hope this answers your question.

**Comment by neolit123** [comment]
/kind support
/remove-sig network
/sig cluster-lifecycle

1.18 was dropped from support a long time ago.
i recall a kubeadm bug where the node where "kubeadm init" was called got the embedded cert/key in kubelet.conf instead of the links to files.

this was fixed a long time ago in kubeadm.
please fix it manually in your cluster by editing the kubelet.conf.
and try to upgrade to >=1.25.

/close

**Comment by stricklandye** [comment]
@pranav-pandey0804 Sorry for not replying in time. I just recently took over the management of this cluster so I am not clear about it's past :D. Maybe other people makes some mistakes. In addition to answer my questions, you have also help me figure out knowledge related to k8s cluster certificates. 
Thank you for your patience.

**Comment by stricklandye** [comment]
@neolit123 I see. In the past I thought the bug in `kubeadm init` should have been fixed in 1.18 and now I know what's really going on here. Thank you very much.
